File: frameworks/go/gorilla-mux/Implementation/middleware.go
package implementation

import (
	"net/http"

	"github.com/gorilla/mux"
)

// middlewareRoutes put no-op layers in front of the handler. A mux middleware belongs to a
// router and runs for every request one of the router's routes matches, so each layered route
// is a subrouter of its own path.
func middlewareRoutes(r *mux.Router, p *Payloads) {
	small := func(w http.ResponseWriter, r *http.Request) { respond(w, http.StatusOK, &p.Small) }

	r.HandleFunc("/middleware/none", small).Methods(http.MethodGet)

	// rb:handler middleware.four
	four := r.Path("/middleware/four").Subrouter()
	four.Use(layered(4)...)
	four.Methods(http.MethodGet).HandlerFunc(small)
	// rb:end

	// rb:handler middleware.sixteen
	sixteen := r.Path("/middleware/sixteen").Subrouter()
	sixteen.Use(layered(16)...)
	sixteen.Methods(http.MethodGet).HandlerFunc(small)
	// rb:end
}

// rb:wiring middleware.*
// noop is one layer: it calls the next handler and does nothing else.
func noop(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { next.ServeHTTP(w, r) })
}

// layered is count no-op layers. mux wraps a router's layers around the matched handler on
// every request.
func layered(count int) []mux.MiddlewareFunc {
	layers := make([]mux.MiddlewareFunc, count)
	for i := range layers {
		layers[i] = noop
	}
	return layers
}

// rb:end

File: frameworks/python/litestar/Implementation/routes/middleware.py
from litestar import Router, get
from litestar.enums import ScopeType
# rb:wiring middleware.*
from litestar.middleware import ASGIMiddleware
from litestar.types import ASGIApp, Receive, Scope, Send

from payloads import Payload, Payloads


# rb:wiring middleware.*
class Noop(ASGIMiddleware):
    """One layer: it calls the next and does nothing else. ASGIMiddleware is the base Litestar's
    documentation recommends for a middleware of one's own."""

    scopes = (ScopeType.HTTP,)

    async def handle(self, scope: Scope, receive: Receive, send: Send, next_app: ASGIApp) -> None:
        await next_app(scope, receive, send)


def layers(count: int) -> list[Noop]:
    return [Noop() for _ in range(count)]
# rb:end


def router(p: Payloads) -> Router:
    """middleware: no-op layers on the route, in front of the handler."""

    @get("/middleware/none")
    async def none() -> Payload:
        return p.small

    @get("/middleware/four", middleware=layers(4))
    async def four() -> Payload:
        return p.small

    @get("/middleware/sixteen", middleware=layers(16))
    async def sixteen() -> Payload:
        return p.small

    return Router(path="/", route_handlers=[none, four, sixteen])
